iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/softwareurl/HttpContentRetriever.java

```diff
@@ -1,77 +1,207 @@
 
-
     private static final long serialVersionUID = -6879262115292175343L;
     
     private static final Logger log = Logger.getLogger(HttpContentRetriever.class);
+    
+    private static final String HEADER_LOCATION = "Location";
+    
+    private int connectionTimeout;
+    
+    private int readTimeout;
+    
+    private int maxPageContentLength;
+    
+    private long throttleSleepTime;
+    
+    private int maxRetriesCount;
+    
+    // to be reinitialized after deserialization
+    private transient CloseableHttpClient httpClient;
 
     
+    // ----------------------------------------- CONSTRUCTORS ---------------------------------------
+    
+    
+    public HttpContentRetriever(int connectionTimeout, int readTimeout, int maxPageContentLength,
+            long throttleSleepTime, int maxRetriesCount) {
+        initialize(connectionTimeout, readTimeout, maxPageContentLength, throttleSleepTime, maxRetriesCount);
+    }
+
+    
+    // ----------------------------------------- LOGIC ----------------------------------------------
+    
     @Override
-    public ContentRetrieverResponse retrieveUrlContent(CharSequence url, int connectionTimeout, int readTimeout,
-            int maxPageContentLength) {
+    public ContentRetrieverResponse retrieveUrlContent(CharSequence url) {
+
         long startTime = System.currentTimeMillis();
-        String currentUrl = url.toString();
-        
+
+        log.info("starting content retrieval for url: " + url);
         try {
-            log.info("starting content retrieval for url: " + currentUrl);
-            
-            HttpURLConnection conn = (HttpURLConnection) new URL(currentUrl).openConnection();
-            conn.setReadTimeout(readTimeout);
-            conn.setConnectTimeout(connectionTimeout);
+            return retrieveUrlContent(url.toString(), 0);
+        } catch (Exception e) {
+            log.error("content retrieval failed for url: " + url, e);
+            return new ContentRetrieverResponse(e);
+        } finally {
+            log.info("finished content retrieval for url: " + url + " in " + (System.currentTimeMillis() - startTime)
+                    + " ms");
+        }
+    }
+    
+    // ----------------------------------------- PRIVATE ------------------------------------------------
+    
+    /**
+     * Retrieves web page content from given url.
+     * 
+     * This method is recursive and requires response entity to be consumed in order
+     * not to hit the ConnectionPoolTimeoutException when connecting the same host
+     * more than 2 times within recursion (e.g. when reattepmting).
+     */
+    private ContentRetrieverResponse retrieveUrlContent(String currentUrl, int retryCount) throws MalformedURLException, IOException, InterruptedException {
+        
+        if (retryCount > maxRetriesCount) {
+            String message = String.format("number of maximum retries exceeded: '%d' for url: %s", maxRetriesCount, currentUrl);
```

Comment:
Well, the thing is `{}` placeholders work for slf4j but here we reuse the string when constructing the exception object.

This my major doubt about replacing log4j with slf4j because it changes the way we refer to placeholders whenever log4j + `String.format()` combo was used. Not mentioning cases where log4j `log()` method was used with `Throwable` parameter - we would need to be careful when introducing list of params following the `{}` placeholders not to swallow `Throwable` as one of the parameters...

Reply:
You're right, I've missed that `message` is reused.

SLF4J has the same methods for logging exceptions as Log4J, that is, `log.level(String msg, Throwable t)`.